### PR TITLE
HY & AG - Personal schedule dropdown component

### LINKED ISSUE #3: Create a dropdown component for selecting a personal schedule
Later on, we want users to be able to add a course to any personal schedule. For that we, need a nice dropdown component that lists all of the current personal schedules.

Using `frontend/src/main/components/Quarters/SingleQuarterDropdown.js` as a model, create a component that takes a list of personal schedule objects as a parameter (i.e. a prop), and produces a dropdown that has a list of personal schedules.

Each schedule in the dropdown should have as it's value the id of the personal schedule, and as the option that displays, the quarter (in QXX format, e.g. `F22`) followed by the first 15 characters of the name of the personal schedule.

Note that another issue is making sure that the names will not be longer than 15 characters in the future, but you don't need to wait for that to be done. For now, just display the first 15 characters of whatever you get.

### PR DESCRIPTION
**Pull request by HansonYu23**
Created a Personal Schedule dropdown list that shows all personal schedules in the format "id - QYY - name". Personal Schedules can be selected from this dropdown list and the state of the form should reflect the Personal Schedule selected. All frontend tests and mutation tests are passing on local machines. 

Storybook Link: https://ucsb-cs156-f22.github.io/f22-7pm-courses-docs-qa/storybook-qa/PersonalScheduleDropdown/?path=/docs/components-personalschedules-singlepersonalscheduledropdown--one-personal-schedule

Resolves  #3

### CODE REVIEW

**Comment by pconrad** [comment]
@HansonYu23 Typically, it's only the staff (instructors, TAs, LAs) that add the labels for points.  But your suggestion that this may be a 20 points issue is noted :-)

**Review by pconrad — CHANGES_REQUESTED** [review]
Please add a story file for this; when adding new components that aren't part of the app yet, that's the only way to really visualize them.

You can look the other files under stories for examples.  There are two styles: .mdx and .js each with pros/cons.  The .mdx format is more expressive, but can be tougher to debug.  The .js format is a bit more constrained but easier to debug.  Either is fine.

**Comment by HansonYu23** [comment]
Added story file

**Comment by pconrad** [comment]
> Added story file

You've added a story file, it's true.  But you didn't link to the deployed storybook entry.

And, the deployed storybook entry doesn't work:

https://ucsb-cs156-f22.github.io/f22-7pm-courses-docs-qa/storybook-qa/PersonalScheduleDropdown/?path=/docs/components-personalschedules-singlepersonalscheduledropdown--one-personal-schedule

At least all I saw was a blank screen. What I'm supposed to be able to see there is a demonstration of how the feature works.

Looking at the console output, I see errors:

```
TypeError: personalSchedules.map is not a function
```

image

This is the line of code that I think is responsible.   I'll leave it as an exercise to you to see if I'm correct, and what needs to be changed. (I think I know, but I need to leave some of the work for you.)

```
    personalSchedules: personalSchedulesFixtures.onePersonalSchedule
```

**Comment by pconrad** [comment]
Here's the correct storybook entry:

https://ucsb-cs156-f22.github.io/f22-7pm-courses-docs-qa/storybook-qa/PersonalScheduleDropdown/?path=/docs/components-personalschedules-singlepersonalscheduledropdown--one-personal-schedule

File: frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  const personalSchedule = jest.fn();
+  const setPersonalSchedule = jest.fn();
+
+  test("renders without crashing on one personalSchedule", () => {
+    render(
+      <SinglePersonalScheduleDropdown
+        personalSchedules={personalSchedulesFixtures.onePersonalScheduleArray}
+        personalSchedule={personalSchedulesFixtures.onePersonalScheduleArray}
+        setPersonalSchedule={setPersonalSchedule}
+        controlId="ssd1"
```
Comment (pconrad) [inline frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js]: psd1 ?

File: frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    );
+  });
+
+  test("renders without crashing on three personalSchedules", () => {
+    render(
+      <SinglePersonalScheduleDropdown
+        personalSchedules={personalSchedulesFixtures.threePersonalSchedules}
+        personalSchedule={personalSchedule}
+        setPersonalSchedule={setPersonalSchedule}
+        controlId="ssd1"
```
Comment (pconrad) [inline frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js]: psd1 ?

File: frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    );
+  });
+
+  test("when I select an object, the value changes", async () => {
+    render(
+      <SinglePersonalScheduleDropdown
+        personalSchedules={personalSchedulesFixtures.threePersonalSchedules}
+        personalSchedule={personalSchedule}
+        setPersonalSchedule={setPersonalSchedule}
+        controlId="ssd1"
```
Comment (pconrad) [inline frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js]: psd1?

File: frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  });
+
+  test("if I pass a non-null onChange, it gets called when the value changes", async () => {
+    const onChange = jest.fn();
+    render(
+      <SinglePersonalScheduleDropdown
+        personalSchedules={personalSchedulesFixtures.threePersonalSchedules}
+        personalSchedule={personalSchedule}
+        setPersonalSchedule={setPersonalSchedule}
+        controlId="ssd1"
```
Comment (pconrad) [inline frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js]: psd1 ?

File: frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    expect(event.target.value).toBe("2");
+  });
+
+  test("default label is Personal Schedule", async () => {
+    render(
+      <SinglePersonalScheduleDropdown
+        personalSchedules={personalSchedulesFixtures.threePersonalSchedules}
+        personalSchedule={personalSchedule}
+        setPersonalSchedule={setPersonalSchedule}
+        controlId="ssd1"
```
Comment (pconrad) [inline frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js]: psd1 ?

File: frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    expect(await screen.findByLabelText("Personal Schedule")).toBeInTheDocument();
+  });
+
+  test("keys / testids are set correctly on options", async () => {
+    render(
+      <SinglePersonalScheduleDropdown
+        personalSchedules={personalSchedulesFixtures.threePersonalSchedules}
+        personalSchedule={personalSchedule}
+        setPersonalSchedule={setPersonalSchedule}
+        controlId="ssd1"
```
Comment (pconrad) [inline frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js]: psd1 ?

File: frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+    const setPersonalScheduleStateSpy = jest.fn();
+    useState.mockImplementation((x) => [x, setPersonalScheduleStateSpy]);
+
+    render(
+      <SinglePersonalScheduleDropdown
+        personalSchedules={personalSchedulesFixtures.threePersonalSchedules}
+        personalSchedule={personalSchedule}
+        setPersonalSchedule={setPersonalSchedule}
+        controlId="ssd1"
```
Comment (pconrad) [inline frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js]: psd1 ?

File: frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
```
Comment (pconrad) [inline frontend/src/tests/components/PersonalSchedules/SinglePersonalScheduleDropdown.test.js]: etc...

**Review by pconrad — CHANGES_REQUESTED** [review]
One small thing... ssd1 is for "Single Schedule Dropdown 1".

It doesn't make a different if it's just in a test, but if someone uses that code in the test as an example when they go to implement this on a page, it would interfere with the Local Storage Key for the single schedule dropdown.

So it would be good to search/replace ssd1 with something like psd1

**Comment by HansonYu23** [comment]
Resolved all requested changes, storybook entry should now work as intended.

**Review by pconrad — APPROVED** [review]
LGTM

**Pull request merged**